### DISCUSSION TITLE
Next.js 13 (app version) background revalidation and caching dynamic segments

### DISCUSSION THREAD

**Discussion by ceolinwill** [vercel/next.js#43085] [Help] [answered]
I'm trying out the new `app` version. Are requests made to a dynamic route segment (i.e. `/posts/123`) cached by default?

For example:

1. I navigate to `/posts/123` for the first time. It builds the `/app/posts/[postID]/page.js` page/file and it sends a `fetch` request to the database to retrieve the `123` post.
2. I go to `/posts/123` again.

What's the expected behavior? Should that page be cached or should a new request to the database be made?

Reading the [background revalidation docs](https://beta.nextjs.org/docs/data-fetching/revalidating#background-revalidation), I thought the page would be cached and another `fetch` request wouldn't be made until the `revalidate` period expired:

```
When a request is made to a route segment that hasn’t been generated, Next.js will dynamically render the route on the first request. Future requests will serve the static route segments from the cache.
```

However, when I tried out this feature, it looked like the `fetch` request was being made every time I visited the `/posts/123` page - not only on the first time.

Here's the code I tried:

```js
// app/posts/[postID]/page.js
async function getData(id) {
  const res = await fetch(`http://localhost:3004/posts/${id}`, { next: { revalidate: 3600 } });
  return res.json();
}

export default async function Page({ params }) {
  const posts = await getData(params.postID);

  return <div>{post.id} - {post.title}</div>
}

// pages/api/posts/[postID].js
export default function handler(req, res) {
  console.log('>>>> loaded');
  res.status(200).json({ id: 1, title: 'John Doe' });
}
```

I've tried it by running `next build` and `next start`. Is there something wrong here or `fetch` being called on every request is the expected behavior?

**Comment by LouisHaftmann** [comment, marked as answer]
Have you tried adding [generateStaticParams](https://beta.nextjs.org/docs/api-reference/generate-static-params)?

Something like this should work:
```ts
export async function generateStaticParams() {
  return []
}
```

**Reply by ceolinwill** [reply]
But `generateStaticParams` works at build time, doesn't it? 🤔

This means I'd need to build all pages at build time, which isn't what I'm looking for here because the build process would take a long time to complete.

What I'm looking for is for a solution that generates a page on-demand but caches it until revalidation.

**Reply by LouisHaftmann** [reply]
You don't need to return any params in `generateStaticParams`. It's just required for `revalidate` to work with dynamic route segments.

**Reply by ceolinwill** [reply]
I see. I'm going to give it a try. Thanks!

**Reply by ceolinwill** [reply]
I finally managed to test this, @LouisHaftmann. It does work, thanks!
